[DX-1688, TT-13153] endpoint level rate limits (#5409)
TykTechnologies/tyk-docs: tyk-docs/content/basic-config-and-security/control-limit-traffic/rate-limiting.md

{{< youtube n7jbmuWgPsw >}}

## Setting up a Key-Level Per-endpoint rate limit

> If you want to restrict API clients to a certain rate of requests for a specific endpoint of an API, you will also configure the rate limiter via the security policy. However this time you'll assign per-endpoint limits. The allowance that you configure in the policy will be consumed by any requests made to that specific endpoint using a key generated from that policy. Thus, if a policy grants access to an API endpoint with `rate=5 per=60` then three clients using keys generated from that policy will each independently be able to make 5 requests in any 60 second period to that particular endpont of the API before receiving the `HTTP 429 Too Many Requests` error. Endpoints of the API that doesn't have endpoint level rate limits configured would be using the rate limits inherited to the API either via per-API rate limit or key level global rate limits based on the configurations.

Reviewer: ```suggestion
If you want to restrict the request rate for specific API clients to a specific endpoint of an API, you should use the security policy to assign per-endpoint rate limits. The allowance that you configure in the policy will be consumed by any requests made to that specific endpoint using a key generated from that policy. Each key will have its own allowance, for example: if a policy grants access to an API endpoint with `rate=5 per=60` then three clients using keys generated from that policy will each independently be able to make 5 requests in any 60 second period to that particular endpont of the API before receiving the `HTTP 429 Too Many Requests` error.

Any API endpoints that don't have endpoint-level rate limits configured will inherit the key-level per-API or key-level global rate limit depending upon which is configured.
```

{{< note success >}}
**Note**  

>  Following assumptions are made for the aforementioned setup.

Reviewer: ```suggestion
The following assumptions are made:
```

>  - The APIs being protected with a rate limit are using our [Authentication token]({{< ref "basic-config-and-security/security/authentication-authorization/bearer-tokens" >}}) Authentication mode and have policies already created. 

Reviewer: ```suggestion
 - the APIs for which rate limits are being set are protected using [authentication token]({{< ref "basic-config-and-security/security/authentication-authorization/bearer-tokens" >}}) and have policies already created
```

Reviewer: This assumption is not valid and has been removed.

>  - Ignore authentication plugin is not turned on for these endpoints.

Reviewer: ```suggestion
 - the [ignore authentication]({{< ref "product-stack/tyk-gateway/middleware/ignore-middleware" >}}) middleware is not turned enabled for these endpoints
```

>  - Path based permissions, if configured grants access to these endpoints

Reviewer: ```suggestion
 - if [path based permissions]({{< ref "getting-started/create-security-policy#path-based-permissions" >}}) are configured for the API then these grant access to these endpoints for keys generated from the policies
```

{{< /note >}}

> You can configure this rate limit from the API Designer in Tyk Dashboard as follows:

Reviewer: ```suggestion
You can configure per-endpoint rate limits from the API Designer in Tyk Dashboard as follows:
```

1. Navigate to the Tyk policy for which you want to set the rate limit
2. Ensure that API that you want to apply rate limits to is selected

> 3. Under **API Access** -> **Set endpoint-level usage limits** click on **Add Rate Lmit** to configure endpoint level rate limits.

Reviewer: ```suggestion
3. Under **API Access** -> **Set endpoint-level usage limits** click on **Add Rate Limit** to configure the rate limit. You will need to provide the rate limit and the endpoint path and method.
```

{{< youtube n7jbmuWgPsw >}}

> ## Setting up a Key-Level Per-endpoint rate limit

Reviewer 2: ```suggestion
## Setting up a key-level per-endpoint rate limit
```

> If you want to restrict the request rate for specific API clients to a specific endpoint of an API, you should use the security policy to assign per-endpoint rate limits. The allowance that you configure in the policy will be consumed by any requests made to that specific endpoint using a key generated from that policy. Each key will have its own allowance, for example: if a policy grants access to an API endpoint with `rate=5 per=60` then three clients using keys generated from that policy will each independently be able to make 5 requests in any 60 second period to that particular endpont of the API before receiving the `HTTP 429 Too Many Requests` error.

Reviewer 2: ```suggestion
If you want to restrict the request rate for specific API clients to a specific endpoint of an API, you should use the security policy to assign per-endpoint rate limits. The allowance that you configure in the policy will be consumed by any requests made to that specific endpoint using a key generated from that policy. Each key will have its own allowance, for example: if a policy grants access to an API endpoint with `rate=5 per=60` then three clients using keys generated from that policy will each independently be able to make 5 requests in any 60 second period to that particular endpoint of the API before receiving the `HTTP 429 Too Many Requests` error.
```

> Any API endpoints that don't have endpoint-level rate limits configured will inherit the key-level per-API or key-level global rate limit depending upon which is configured.

Reviewer 3: @andyo-tyk I think you missed this rephrasing suggestion, do you want to include it?

> Any API endpoints that don't have endpoint-level rate limits configured will inherit the key-level per-API or key-level global rate limit depending upon which is configured.

Reviewer: Yes, looks good to me (though we usually put the HTTP response codes/messages in `code` format.

> If you want to restrict the request rate for specific API clients to a specific endpoint of an API, you should use the security policy to assign per-endpoint rate limits. The allowance that you configure in the policy will be consumed by any requests made to that specific endpoint using a key generated from that policy. Each key will have its own allowance, for example: if a policy grants access to an API endpoint with `rate=5 per=60` then three clients using keys generated from that policy will each independently be able to make 5 requests in any 60 second period to that particular endpoint of the API before receiving the `HTTP 429 Too Many Requests` error.
> 
> Any API endpoints that don't have endpoint-level rate limits configured will inherit the key-level per-API or key-level global rate limit depending upon which is configured.

Reviewer 3: ```suggestion
To restrict the request rate for specific API clients on particular endpoints, you can use the security policy to assign per-endpoint rate limits. These limits are set within the policy and will be #enforced for any requests made to that endpoint by clients using keys generated from that policy.
Each key will have its own independent rate limit allowance. For example, if a policy grants access to an endpoint with a rate limit of 5 requests per 60 seconds, each client with a key from that policy can make 5 requests to the endpoint in any 60-second period. Once the limit is reached, the client will receive an HTTP 429 Too Many Requests error.
If no per-endpoint rate limit is defined, the endpoint will inherit the key-level per-API rate limit or the global rate limit, depending on what is configured.
```

{{< note success >}}
**Note**  
The following assumptions are made:

>  - the [ignore authentication middleware]({{< ref "product-stack/tyk-gateway/middleware/ignore-middleware" >}}) is not enabled for these endpoints

Reviewer 3: ```suggestion
 - The [ignore authentication]({{< ref "product-stack/tyk-gateway/middleware/ignore-middleware" >}}) middleware should not be enabled for the relevant endpoints.
```

>  - the [ignore authentication]({{< ref "product-stack/tyk-gateway/middleware/ignore-middleware" >}}) middleware is not turned enabled for these endpoints

Reviewer 2: ```suggestion
 - the [ignore authentication middleware]({{< ref "product-stack/tyk-gateway/middleware/ignore-middleware" >}}) is not enabled for these endpoints
```
